mrrc/pkgs/indexing.py

```diff
@@ -42,212 +40,217 @@ def get_index_template() -> str:
+
+    for temp_dir in temp_dirs:
+        virtual_folder_delete(temp_dir, top_level, s3_client, bucket,
+                              updated_indexes, delete_indexes)
+
+    html_files = set()
+    if updated_indexes != set():
+        updated_indexes = {os.path.join(top_level, _) for _ in updated_indexes}
+        html_files = index_to_html(updated_indexes, top_level)
+    if delete_indexes != set():
+        for index in set(delete_indexes):
+            delete_indexes.add(os.path.join(os.path.dirname(index), 'index.html'))
+    return delete_indexes, updated_indexes.union(html_files)
+
+
+# e.g path: org/apache/httpcomponents/httpclient/4.5.6, updated_indexes contains every local index
+def virtual_folder_create(
+    path: str, base_dir: str, s3_client: S3Client, bucket: str, updated_indexes: Set[str]
+):
+    item = os.path.basename(path)
+    dir_index = os.path.join(os.path.dirname(path), '.index')
+    local_index_file = os.path.join(base_dir, dir_index)
+    updated_indexes.add(dir_index)
+
+    if os.path.exists(local_index_file):
+        items = load_local_index(local_index_file)
+        if item in items:
+            return
         else:
-            items = load_exist_index(s3_client, bucket, index_file)
-
-        # scan list to find items in same folder, and remove them from .index file
-        for _ in set(deleted_files):
-            if _ in scanned:
-                continue
-            if '/' in _ and os.path.dirname(_) == path:
-                last_path = os.path.basename(_)
-                if last_path in items:
-                    items.remove(last_path)
-                elif last_path+'/' in items:
-                    items.remove(last_path + '/')
-                scanned.add(_)
-                deleted_files.remove(_)
-
-        if path == '' and (d_file in items or d_file + '/' in items):
-            if d_file in items:
-                items.remove(d_file)
-            elif d_file + '/' in items:
-                items.remove(d_file + '/')
-            deleted_files.remove(d_file)
-
-        # if every item has been removed from list, it means the containing folder is empty
-        # thus add the folder to removed files to scann its upper level index
+            # only appends line, no truncate and no overwrite
+            with open(local_index_file, 'a', encoding='utf-8') as f:
+                f.write(item + '/\n')
+    else:
+        items = load_s3_index(s3_client, bucket, dir_index)
         if items == set():
-            delete_index.append(local_index_file)
-            if path != '':
-                deleted_files.add(path)
-            if local_index_file in update_index:
-                update_index.remove(local_index_file)
-        # if there is still items left, save it on local file, and mark it has already been updated
+            with open(local_index_file, 'a+', encoding='utf-8') as f:
+                f.write(item + '/\n')
+            if dir_index != '.index':
```

Comment:
Should this condition happen before line#149? If it is not an .index, why should it be loaded from s3 again?